distod/src/main/scala/com/github/codelionx/distod/actors/worker/Worker.scala

```diff
@@ -27,94 +30,211 @@ object Worker {
-    partitionManager ! LookupAttributes(partitionEventMapper)
 
-    Behaviors.receiveMessagePartial {
-      case WrappedPartitionEvent(AttributesFound(attributes)) =>
-        context.log.trace("Worker ready to process candidates: Master={}, Attributes={}", master, attributes)
+  private object StopBehaviorInterceptor extends WorkerStopInterceptor {
+
+    override def dispatchWorkFromMaster(): Unit =
+      if (!stopped) {
         master ! DispatchWork(context.self)
-        behavior(attributes)
+        openDispatches += 1
+      } else {
+        context.log.trace("Preventing the request for more work, because worker is in shut down phase")
+      }
 
-      case Stop =>
-        Behaviors.stopped
+    override protected def cancelAll(): Unit = {
+      for (job <- splitJobs.keys) {
+        master ! CancelWork(job, JobType.Split)
+      }
+      for (job <- swapJobs.keys) {
+        master ! CancelWork(job, JobType.Swap)
+      }
     }
-  }
-
-  private def behavior(attributes: Seq[Int], stopped: Boolean = false): Behavior[Command] = Behaviors.receiveMessage {
-    case CheckSplitCandidates(candidateId, splitCandidates) =>
-      context.log.debug("Checking split candidates of node {}", candidateId)
 
-      def handleResults(removedSplitCandidates: CandidateSet = CandidateSet.empty): Behavior[Command] = {
-        // notify master of result
-        context.log.debug("Sending results of ({}, Split) to master at {}", candidateId, master)
-        master ! SplitCandidatesChecked(candidateId, removedSplitCandidates)
-
-        // ready to work on next node:
-        next(attributes, stopped)
-      }
+    override protected def cancel(candidateId: CandidateSet, jobType: JobType.JobType): Unit =
+      master ! CancelWork(candidateId, jobType)
+  }
 
-      if (splitCandidates.nonEmpty) {
-        SplitCandidateValidationBehavior(workerContext, attributes, candidateId, splitCandidates)(
-          removedSplitCandidates => handleResults(removedSplitCandidates)
-        )
-      } else {
-        handleResults()
-      }
 
-    case CheckSwapCandidates(candidateId, swapCandidates) =>
-      context.log.debug("Checking swap candidates of node {}", candidateId)
+  def start(): Behavior[Command] = {
+    context.log.trace(
+      "Worker ready to process candidates: Master={}, Attributes={}, requesting {} jobs",
+      master,
+      attributes,
+      settings.concurrentWorkerJobs
+    )
+    for (_ <- 0 until settings.concurrentWorkerJobs) {
+      StopBehaviorInterceptor.dispatchWorkFromMaster()
+    }
+    behavior()
+  }
 
-      def handleResults(removedSwapCandidates: Seq[(Int, Int)] = Seq.empty): Behavior[Command] = {
-        // notify master of result
-        context.log.debug("Sending results of ({}, Swap) to master at {}", candidateId, master)
-        master ! SwapCandidatesChecked(candidateId, removedSwapCandidates)
+  private def behavior(): Behavior[Command] = Behaviors.intercept(() => StopBehaviorInterceptor)(
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method is longer than 50 lines](https://app.codacy.com/manual/CodeLionX/distod/pullRequest?prid=5048550)